lib/mayaUsd/fileio/primUpdaterManager.cpp

```diff
@@ -1724,4 +1747,270 @@ bool PrimUpdaterManager::readPullInformation(const MDagPath& dagPath, Ufe::Path&
+                }
+            }
+            if (!foundChild) {
+                // Following a subtree invalidate, if none of the now-valid
+                // children appear in the trie, means that we've switched to a
+                // different variant, and everything below that path should be
+                // hidden.
+                auto ancestorNode = _pulledPrims.node(op.path);
+                if (ancestorNode) {
+                    recursiveSetVisibility(ancestorNode, false);
+                }
+            }
+        }
+    } break;
+    }
+}
+
+/* static */
+bool PrimUpdaterManager::OrphanedNodesManager::setVisibilityPlug(
+    const Ufe::TrieNode<PullVariantInfo>::Ptr& trieNode,
+    bool                                       visibility)
+{
+    TF_VERIFY(trieNode->hasData());
+    const auto& pullParentPath = trieNode->data().dagPath;
+    MFnDagNode  fn(pullParentPath);
+    auto        visibilityPlug = fn.findPlug("visibility", /* tryNetworked */ true);
+    return (visibilityPlug.setBool(visibility) == MS::kSuccess);
+}
+
+/* static */
+void PrimUpdaterManager::OrphanedNodesManager::recursiveSetVisibility(
+    const Ufe::TrieNode<PullVariantInfo>::Ptr& trieNode,
+    bool                                       visibility)
+{
+    // We know in our case that a trie node with data can't have children,
+    // since descendants of a pulled prim can't be pulled.
+    if (trieNode->hasData()) {
+        TF_VERIFY(trieNode->empty());
+        TF_VERIFY(setVisibilityPlug(trieNode, visibility));
+    } else {
+        auto childrenComponents = trieNode->childrenComponents();
+        for (const auto& c : childrenComponents) {
+            recursiveSetVisibility((*trieNode)[c], visibility);
+        }
+    }
+}
+
+/* static */
+void PrimUpdaterManager::OrphanedNodesManager::recursiveSwitch(
+    const Ufe::TrieNode<PullVariantInfo>::Ptr& trieNode,
+    const Ufe::Path&                           ufePath)
+{
+    // We know in our case that a trie node with data can't have children,
+    // since descendants of a pulled prim can't be pulled.  A trie node with
+    // data is one that's been pulled.
+    if (trieNode->hasData()) {
+        TF_VERIFY(trieNode->empty());
+
+        auto pulledNode = std::dynamic_pointer_cast<MayaUsd::ufe::UsdSceneItem>(
+            Ufe::Hierarchy::createItem(ufePath));
+        if (!TF_VERIFY(pulledNode)) {
+            return;
+        }
+
+        // If the variant set configuration of the pulled node and the current
+        // tree state don't match, the pulled node must be made invisible.
+        // Inactivation must not be considered, as the USD pulled node is made
+        // inactive on pull, to avoid rendering it.
+        const bool variantSetsMatch
+            = (trieNode->data().variantSetDescriptors == variantSetDescriptors(ufePath.pop()));
+        const bool visibility = (pulledNode && variantSetsMatch);
+        TF_VERIFY(setVisibilityPlug(trieNode, visibility));
```

Comment:
Again, there maybe multiple causes of invisiblity, so we should check that all have been restored before making a node visible.